scipy/_lib/_array_api.py

```diff
@@ -177,42 +177,48 @@ def is_torch(xp):
     return xp.__name__ == 'scipy._lib.array_api_compat.array_api_compat.torch'
 
 
-def assert_equal(actual, desired, err_msg='', xp=None):
+def xp_assert_equal(actual, desired, err_msg='', xp=None):
     if xp is None:
         xp = array_namespace(actual)
     if is_cupy(xp):
+        assert actual.dtype == desired.dtype
         return xp.testing.assert_array_equal(actual, desired, err_msg=err_msg)
     elif is_torch(xp):
         # PyTorch recommends using `rtol=0, atol=0` like this
         # to test for exact equality
         return xp.testing.assert_close(actual, desired, rtol=0, atol=0,
                                        msg=err_msg)
```

Comment:
```suggestion
                                       msg=err_msg, equal_nan=True)
```
I think this will fix https://github.com/scipy/scipy/pull/18930#discussion_r1315898607